addons/website/models/website.py

```diff
@@ -827,6 +844,36 @@ def get_base_url(self):
         self.ensure_one()
         return self._get_http_domain() or super(BaseModel, self).get_base_url()
 
+    def _get_canonical_url_localized(self, lang, canonical_params, lang_in_url=True):
+        """Returns the canonical URL for the current request, with or without
+        the language path depending on `lang_in_url`, and with translatable
+        elements appropriately translated in `lang`.
+
+        If `request.endpoint` is not true, returns the current `path` instead.
+
+        `url_quote_plus` is applied on the returned path.
+        """
+        self.ensure_one()
+        if request.endpoint:
+            router = request.httprequest.app.get_db_router(request.db).bind('')
+            arguments = dict(request.endpoint_arguments)
+            for key, val in list(arguments.items()):
+                if isinstance(val, models.BaseModel):
+                    arguments[key] = val.with_context(lang=lang.url_code)
+            path = router.build(request.endpoint, arguments)
+        else:
+            # The build method returns a quoted URL so convert in this case for consistency.
+            path = urls.url_quote_plus(request.httprequest.path, safe='/')
+        lang_path = ('/' + lang.url_code) if lang_in_url else ''
+        canonical_query_string = '?%s' % urls.url_encode(canonical_params) if canonical_params else ''
+        return self.get_base_url() + lang_path + path + canonical_query_string
```

Comment:
as discuss previously, can we split canonical job, and absolute domain... 
E.g. 
`_get_canonical_url`  that `return lang_path + path + canonical_query_string`
and 
`_get_absolute_canoncial_url` that `return  self.get_base_url() + _get_canonical_url`